staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler_test.go

```diff
@@ -638,6 +648,217 @@ func testHandlerConversion(t *testing.T, enableWatchCache bool) {
 	}
 }
 
+func TestDecoder(t *testing.T) {
+	multiVersion := `
+	{
+	"apiVersion": "stable.example.com/v1beta1",
+	"kind": "MultiVersion",
+	"metadata": {
+		"name": "my-mv"
+	},
+	"num": 1,
+	"num": 2,
+	"unknown": "foo"
+	}
+	`
+	multiVersionYAML := `
+apiVersion: stable.example.com/v1beta1
+kind: MultiVersion
+metadata:
+  name: my-mv
+num: 1
+num: 2
+unknown: foo`
+
+	decoded := &unstructured.Unstructured{Object: map[string]interface{}{}}
+	decoded.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   "stable.example.com",
+		Version: "v1beta1",
+		Kind:    "MultiVersion",
+	})
+	decoded.SetName("my-mv")
+	decoded.SetGeneration(1)
+	unstructured.SetNestedField(decoded.Object, int64(2), "num")
+	unstructured.SetNestedField(decoded.Object, nil, "metadata", "creationTimestamp")
+
+	decodedPreserveUnknown := decoded.DeepCopy()
+	unstructured.SetNestedField(decodedPreserveUnknown.Object, "foo", "unknown")
+
+	testcases := []struct {
+		name                  string
+		body                  string
+		yaml                  bool
+		strictDecoding        bool
+		preserveUnknownFields bool
+		crd                   *v1.CustomResourceDefinition
+		expectedObj           *unstructured.Unstructured
+		expectedErr           error
+	}{
+		{
+			name:           "strict-decoding",
+			body:           multiVersion,
```

Comment:
In general, I'd rather see exactly the inputs hard-coded here rather than have to scroll back and forth to see what is referring to what. The test would be more verbose, longer, and repeating itself a lot more, but it'd be also much easier to review, inspect and debug IMO.